src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpMetricExporter.cs

```diff
@@ -69,56 +88,52 @@ public override ExportResult Export(in Batch<Metric> metrics)
 
             // [Performance] Need to allow for a message with partial data points but several metrics so we can still batch metrics in a single message
             //               and minimize the amount of requests to transmit all the data.
-            var perMetricBatches = new List<OtlpMetrics.Metric>[metrics.Count];
-            var meterNames = new Metric[metrics.Count]; // [abeaulieu] Would be nice to be able to do that without copying the metrics.
+            var perMetricBatches = ArrayPool<List<OtlpMetrics.Metric>>.Shared.Rent((int)metrics.Count);
+            var meterNames = ArrayPool<Metric>.Shared.Rent((int)metrics.Count);
 
             var i = 0;
             foreach (var metric in metrics)
             {
-                perMetricBatches[i] = metric.ToBatchedOtlpMetric().ToList();
+                perMetricBatches[i] = metric.ToOtlpMetric(this.dataPointBatchSize, this.dataPointbatchingEnabled).ToList();
                 meterNames[i] = metric;
                 ++i;
             }
 
             // [abeaulieu] Now loop through all the metrics until we've sent all the messages to send
-            var done = false;
-            var num_msg = 0;
-            var wire_size = 0;
-            while (!done)
+            OtlpCollector.ExportMetricsServiceRequest request;
+            bool mustWait = false;
+            do
             {
-                var request = new OtlpCollector.ExportMetricsServiceRequest();
-
-                // Try to make progress on every metric that still has data to send.
-                for (i = 0; i < perMetricBatches.Length; ++i)
+                request = null;
+                for (i = 0; i < metrics.Count; ++i)
                 {
-                    done = true; // If all metrics have flushed their data, done will remain true and the loop exits.
-
                     var messages = perMetricBatches[i];
                     if (messages.Count > 0)
                     {
+                        request ??= new OtlpCollector.ExportMetricsServiceRequest();
                         var m = messages[messages.Count - 1]; // Take the last one so there's no need to list copy to shift the list left.
                         messages.RemoveAt(messages.Count - 1); // [Perf]: Use a queue or just index manipulation to avoid memory shenanigans.
                         request.AddMetrics(this.ProcessResource, m, meterNames[i]);
-                        done = false; // Not done until every metric has nothing to send.
-                        num_msg++;
                     }
                 }
 
-                try
+                if (perMetricBatches[0].Count > 0) // We only wait if we have remaining messages to be sent
```

Comment:
I think this can be set at line 101 if `messages.Count > 0` for any of the `perMetricBatches`, since the loop will grab one batch per metric.

I also just noticed that we are not tallying up the number of points being added to the message here, so if the limit  is 500 but we have 3 metrics with 500 data points each, we'd end up with a single message containing 1500 data points.

Reply:
We probably want an extra unit test to make sure that we don't accidentally batch more than BatchSize across multiple metrics.

e.g.:

batchSize = 1000
m1 has 1000 datapoints
m2 has 500 datapoints

This should result in 2 messages. (m1 should export its 1000 datapoints as a single export request)
m2's pending data points should be in their own export request.

With the current code, this test should fail as it would cause a single export request to be sent.